torch/csrc/jit/script/script_type_parser.cpp

```diff
@@ -152,6 +152,39 @@ c10::optional<std::string> ScriptTypeParser::parseBaseTypeName(
   return at::nullopt;
 }
 
+// Parse NamedTuple following the grammar like below:
+// NamedTuple('Employee', [('name', str), ('id', int)])
+// See also: https://docs.python.org/3/library/typing.html#typing.NamedTuple
+TypePtr parseNamedTuple(const Expr& expr) {
+  auto apply = Apply(expr);
+  auto callee = apply.callee();
+  if (callee.kind() != TK_VAR || Var(callee).name().name() != "NamedTuple") {
+    throw ErrorReport(callee.range()) << "Unknown type expression";
+  }
+  auto inputs = apply.inputs();
+  if (inputs.size() != 2 || inputs[0].kind() != TK_STRINGLITERAL ||
+      inputs[1].kind() != TK_LIST_LITERAL)
+  {
+    throw ErrorReport(inputs.range()) << "Bad NamedTuple construct";
```

Comment:
Umm. So I won't ask you to make these proper error messages (though I really should), but at the very least can we give distinct messages for each case, so that it is easier for us to debug what case we're in if someone reports an error?

Reply:
@ezyang I have improved on this part